base/dcp_receiver.go

```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
 
 func (nph NoPasswordAuthHandler) GetCredentials() (username string, password string, bucketname string) {
 	_, _, bucketname = nph.Handler.GetCredentials()
+
 	return "", "", bucketname
 }
 
+// getHighSeqMetadata returns metadata to feed into a DCP client based on the last sequence numbers stored in memory
+func getHighSeqMetadata(bucket Bucket) ([]DCPMetadata, error) {
+	store, ok := AsCouchbaseStore(bucket)
+	if !ok {
+		return nil, errors.New("DCP Client requires bucket to be CouchbaseStore")
+	}
+
+	numVbuckets, err := store.GetMaxVbno()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to determine maxVbNo when creating DCP client: %w", err)
+	}
+
+	_, highSeqNos, statsErr := store.GetStatsVbSeqno(numVbuckets, true)
+	if statsErr != nil {
+		return nil, fmt.Errorf("Unable to obtain high seqnos for one-shot DCP feed: %w", statsErr)
+	}
+
+	metadata := make([]DCPMetadata, numVbuckets)
+	for vbNo := uint16(0); vbNo < numVbuckets; vbNo++ {
+		// TODO: should we/can we initialize VbUUID or other fields?
+		metadata[vbNo].FailoverEntries = make([]gocbcore.FailoverEntry, 0)
+		metadata[vbNo].StartSeqNo = gocbcore.SeqNo(highSeqNos[vbNo])
+		metadata[vbNo].EndSeqNo = gocbcore.SeqNo(uint64(0xFFFFFFFFFFFFFFFF))
+	}
+	return metadata, nil
+}
+
+// StartGOCB2DCPFeed starts a DCP Feed.
+func StartGOCB2DCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
+	metadata, err := getHighSeqMetadata(bucket)
+	if err != nil {
+		return err
+	}
+	bucketName := spec.BucketName
+	// TODO: does this need to be more unique, if two sync gateways connect to same server, will this reject the second DCP feed?
+	feedName := fmt.Sprintf("%s-%s:%s", args.ID, ProductAPIVersion, bucketName)
```

Comment:
Yes, feed IDs must be unique or kv will terminate the connection of the existing client

We use `GenerateDcpStreamName` elsewhere to include a UUID in the name

Reply:
OTOH, if this is for the caching feed / community import, will we expect to see multiple SGWs connecting?

Reply:
Changed to use common function.